### PR TITLE
[CALCITE-5708] Change SUBSTRING result if either of parameters is NULL

### PR DESCRIPTION
**Pull request by zstan**

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java [hunk 1]
```diff
@@ -1042,6 +1042,10 @@ void testDyadicCollateOperator() {
         .columnType("VARCHAR(1) NOT NULL");
     expr("substring('a', 1, '3')")
         .columnType("VARCHAR(1) NOT NULL");
+
+    expr("SUBSTRING(NULL FROM 1 FOR 2)").ok();
```
Comment (herunkang2018) [inline core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:1046]: this test case is not effected by this issue, would you like to add a comment to clarify it?

File: core/src/test/resources/sql/functions.iq [hunk 2]
```diff
@@ -464,4 +464,19 @@ limit 3;
 
 !ok
 
+# SUBSTRING
+-- returns 'null'
+select SUBSTRING(NULL FROM 1 FOR 2);
```
Comment (herunkang2018) [inline core/src/test/resources/sql/functions.iq:469]: can these test cases in quidem file be put together? I'm not sure if this works for later test cases

Comment (zstan) [reply core/src/test/resources/sql/functions.iq:469]: yes, they can.

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java (hunk 1 above)
Comment (zstan) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:1046]: I just expand the tests coverage.

Comment (herunkang2018) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:1046]: I know it, if you add comment to clarify it, it would be better I think.

Comment (zstan) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:1046]: @herunkang2018 fixed, can you make a review plz ? I update description and starting code point in appropriate issue.

**Comment by zstan** [comment]
@herunkang2018 can you make a review plz ? I fill additional description in issue.

**Comment by herunkang2018** [comment]
@zstan I leave two comments in Jira, after that, I will review this PR again.

**Comment by zstan** [comment]
@herunkang2018 thanks ! i check 
```
  @Test
  void testStrcmpFunc() {
    final SqlOperatorFixture f = fixture()
        .setFor(SqlLibraryOperators.STRCMP)
        .withLibrary(SqlLibrary.MYSQL);

    f.checkNull("STRCMP(null, null)");
```
and what kind of runtine it generates for null literals
`EnumerableCalc#implement`
and i check my PR change for SUBSTRING and as for me it looks the same (for now) i.e. did the same as from java doc :

```
If any of the other operands are nullable the
   * returned type will also be nullable. First Arg must be of string type.
   */
  public static final SqlReturnTypeInference ARG0_NULLABLE_VARYING =
      ARG0.andThen(SqlTypeTransforms.TO_NULLABLE)
          .andThen(SqlTypeTransforms.TO_VARYING);
```
i.e. for null literals appropriate runtime implementation will never been called, seems it correct.

**Comment by zstan** [comment]
@herunkang2018 kindly remind help with review if possible, thanks !

File: core/src/main/java/org/apache/calcite/sql/type/SqlTypeFamily.java [hunk 3]
```diff
@@ -244,7 +244,7 @@ public Collection<SqlTypeName> getTypeNames() {
     case EXACT_NUMERIC:
       return SqlTypeUtil.getMaxPrecisionScaleDecimal(factory);
     case INTEGER:
-      return factory.createSqlType(SqlTypeName.BIGINT);
+      return factory.createSqlType(SqlTypeName.INTEGER);
```
Comment (gianm) [inline core/src/main/java/org/apache/calcite/sql/type/SqlTypeFamily.java:247]: Why this change? I would think that `BIGINT` is a good default concrete type for the family `INTEGER`, since it's the least restrictive in the family. Seems like this may have unwanted effects, if it causes inference for the `INTEGER` family to use a too-narrow type.

File: core/src/test/resources/sql/functions.iq [hunk 4]
```diff
@@ -938,5 +938,19 @@ table(AUX.TBLFUN_IDENTITY(3)) as t3(v);
 
 !ok
 
+# SUBSTRING
```
Comment (gianm) [inline core/src/test/resources/sql/functions.iq:941]: I suggest including a test for the 2-arg form too. Like:

```
select SUBSTRING('text' FROM NULL)
```

It should return null.

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java [hunk 5]
```diff
@@ -1040,6 +1040,11 @@ void testDyadicCollateOperator() {
         .columnType("VARCHAR(1) NOT NULL");
     expr("substring('a', 1, '3')")
         .columnType("VARCHAR(1) NOT NULL");
+
+    // Correctly processed null string and params.
```
Comment (gianm) [inline core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:1044]: I suggest including a test for the 2-arg form too. Like:

```
expr("SUBSTRING('text' FROM NULL)").ok();
```

Comment (zstan) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:1044]: done

File: core/src/test/resources/sql/functions.iq (hunk 4 above)
Comment (zstan) [reply core/src/test/resources/sql/functions.iq:941]: done

File: core/src/main/java/org/apache/calcite/sql/type/SqlTypeFamily.java (hunk 3 above)
Comment (zstan) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeFamily.java]: yep, thanks, seems it redundant change

File: core/src/test/resources/sql/functions.iq [hunk 6]
```diff
@@ -938,5 +938,20 @@ table(AUX.TBLFUN_IDENTITY(3)) as t3(v);
...
+-- returns 'null'
+select SUBSTRING(NULL FROM 1 FOR 2);
+select SUBSTRING('text' FROM 1 FOR NULL);
+select SUBSTRING('text' FROM NULL FOR 2);
+select SUBSTRING(s from i for l) from (values ('abc', null, 2)) as t (s, i, l);
```
Comment (herunkang2018) [inline core/src/test/resources/sql/functions.iq:946]: minor comment, better to unify the usage of uppercase and lowercase. You could follow the test cases above.

**Comment by herunkang2018** [comment]
@gianm Thanks for your review, I will review again in 24h.

File: core/src/test/resources/sql/functions.iq (hunk 6 above)
Comment (zstan) [reply core/src/test/resources/sql/functions.iq]: done

**Comment by herunkang2018** [comment]
@zstan @gianm LGTM

**Review by herunkang2018 — APPROVED** [review]

**Comment by zstan** [comment]
thanks a lot, i will squash all commits in final PR soon

**Comment by zstan** [comment]
seems it ready for merging, thanks a lot
